language/move-vm/runtime/src/debug.rs

```diff
@@ -0,0 +1,194 @@
+            breakpoints: BTreeSet::new(),
+            should_take_input: true,
+        }
+    }
+
+    pub(crate) fn debug_loop(
+        &mut self,
+        function_desc: &Function,
+        locals: &Locals,
+        pc: u16,
+        instr: &Bytecode,
+        resolver: &Loader,
+        interp: &Interpreter,
+    ) {
+        let instr_string = format!("{:?}", instr);
+        let function_string = function_desc.pretty_string();
+        let breakpoint_hit = self.breakpoints.contains(&function_string)
+            || self
+                .breakpoints
+                .iter()
+                .any(|bp| instr_string[..].starts_with(bp.as_str()));
+
+        if self.should_take_input || breakpoint_hit {
+            self.should_take_input = true;
+            if breakpoint_hit {
+                let bp_match = self
+                    .breakpoints
+                    .iter()
+                    .find(|bp| instr_string.starts_with(bp.as_str()))
+                    .unwrap()
+                    .clone();
+                println!(
+                    "Breakpoint {} hit with instruction {}",
+                    bp_match, instr_string
+                );
+            }
+            println!(
+                "function >> {}\ninstruction >> {:?}\nprogram counter >> {}",
+                function_string, instr, pc
+            );
+            loop {
+                print!("> ");
+                std::io::stdout().flush().unwrap();
+                let mut input = String::new();
+                match io::stdin().read_line(&mut input) {
+                    Ok(_) => match input.parse::<DebugCommand>() {
+                        Err(err) => println!("{}", err),
+                        Ok(command) => match command {
+                            DebugCommand::Step => {
+                                self.should_take_input = true;
+                                break;
+                            }
+                            DebugCommand::Continue => {
+                                self.should_take_input = false;
+                                break;
+                            }
+                            DebugCommand::Breakpoint(breakpoint) => {
+                                self.breakpoints.insert(breakpoint.to_string());
+                            }
+                            DebugCommand::DeleteBreakpoint(breakpoint) => {
+                                self.breakpoints.remove(&breakpoint);
+                            }
+                            DebugCommand::PrintBreakpoints => self
+                                .breakpoints
+                                .iter()
+                                .enumerate()
+                                .for_each(|(i, bp)| println!("[{}] {}", i, bp)),
+                            DebugCommand::PrintStack => {
+                                let mut s = String::new();
+                                let _ = interp.debug_print_stack_trace(&mut s, resolver);
```

Comment:
So here we're calling a function with side effects and silently discarding its result. A little bit disturbed by this but I understand you can't simply propagate the error up. Would it be better we `unwrap` here? I assume it's fine for it to crash in case something goes wrong in `debug_print_stack_trace` as it only happens in a non-prod build.

Reply:
Good spot, I'll add that `unwrap` in!